### ISSUE TITLE
Don't rely on System.IO namespace for FilePath/DirectoryPath

### ISSUE THREAD

**Issue by patriksvensson** [cake-build/cake#2197]

**Comment by jnm2** [comment]
I would have said each of the listed APIs operates on strings without accessing a file system since I have used all of them with nonexistent paths. Did you look into it and see it doing I/O?

**Comment by patriksvensson** [comment]
@jnm2 It does things like path validation which we don't want.

**Comment by patriksvensson** [comment]
@jnm2 Yes. There is no OS-specific validation, but there are things like ACL security checks (i.e. for getting directory names or expanding paths) and path length validation (see https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/path.cs). What we want to do is to completely remove all kinds of validation of paths and leave the validation to the file system implementation.

The end goal here is to remove all constraints for path representations and by doing that it would be easy to add something like a GlobPath in the future (or UNC support). We will also remove any validation that we currently have in the Path base class.

There are of course drawbacks to this, but I think having a lightweight representation will make it easier for us in the future.
